### ISSUE TITLE
BUG: itemsize wrong for date32[day][pyarrow] dtype?

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#57948]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pyarrow as pa
import pandas as pd

pd.ArrowDtype(pa.date32())  # date32[day][pyarrow]
pd.ArrowDtype(pa.date32()).itemsize  # 8
```

### Issue Description

I think it should show `4`? `pa.date32()` is 32 bits, so 4 bytes

### Expected Behavior

pd.ArrowDtype(pa.date32()).itemsize  # 4

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : b033ca94e7ae6e1320c9d65a8163bd0a6049f40a
python                : 3.10.12.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.15.146.1-microsoft-standard-WSL2
Version               : #1 SMP Thu Jan 11 04:09:03 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : C.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.1.0.dev0+761.gb033ca94e7
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 69.1.1
pip                   : 24.0
Cython                : 3.0.8
pytest                : 8.0.2
hypothesis            : 6.98.15
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.1.0
html5lib              : 1.1
pymysql               : 1.4.6
psycopg2              : 2.9.9
jinja2                : 3.1.3
IPython               : 8.22.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : 1.3.8
fastparquet           : 2024.2.0
fsspec                : 2024.2.0
gcsfs                 : 2024.2.0
matplotlib            : 3.8.3
numba                 : 0.59.0
numexpr               : 2.9.0
odfpy                 : None
openpyxl              : 3.1.2
pyarrow               : 15.0.2
pyreadstat            : 1.2.6
python-calamine       : None
pyxlsb                : 1.0.10
s3fs                  : 2024.2.0
scipy                 : 1.12.0
sqlalchemy            : 2.0.27
tables                : 3.9.2
tabulate              : 0.9.0
xarray                : 2024.2.0
xlrd                  : 2.0.1
zstandard             : 0.22.0
tzdata                : 2024.1
qtpy                  : 2.4.1
pyqt5                 : None

</details>

**Comment by MarcoGorelli** [comment]
Not sure what to do here, as there isn't a numpy dtype corresponding to date32

https://github.com/pandas-dev/pandas/blob/77f9d7abee14888447a1f9942f7f6f4cdbcd927b/pandas/core/dtypes/dtypes.py#L2215-L2218

**Comment by jorisvandenbossche** [comment]
PyArrow data types of fixed width have a `bit_width` attribute that could be used here. That does raise for nested types, though at the moment we just return 8 from the numpy object dtype, which also doesn't necessarily makes sense.

```
In [13]: pd.ArrowDtype(pa.list_(pa.int32())).itemsize
Out[13]: 8
```

**Comment by TinusChen** [comment]
The `pd.ArrowDtype` object does not have an `itemsize` attribute. The `itemsize` attribute is typically related to the size of data items in memory, which is valid in NumPy arrays but not applicable to `pd.ArrowDtype`. In Pandas, the `ArrowDtype` object is used to represent types within Arrow data frames, rather than being directly related to memory size.

If you wish to obtain size information for PyArrow types, you should interact directly with PyArrow-related objects or functions. For instance, if you have a PyArrow RecordBatch or Table, you can inspect their schema to obtain type information.

**Comment by TinusChen** [comment]
The confusion here seems to stem from the expectation that `pd.ArrowDtype(pa.date32())`.itemsize should return 4, because `pa.date32()` represents a 32-bit date type. However, the `itemsize` attribute is not a property of `pd.ArrowDtype` but rather of NumPy data types or Pandas categorical types.

The `pa.date32()` type from PyArrow indeed represents a 32-bit date type, which means it can store a date as a 32-bit integer, typically representing the number of days since a specific epoch (e.g., since 1970-01-01). Since it's 32 bits, it occupies 4 bytes (32 bits / 8 bits per byte).

However, when you create a `pd.ArrowDtype` object with `pa.date32()`, you're creating a Pandas representation of the PyArrow type, which is not directly tied to the memory size of individual values but rather to the metadata describing the type.

If you want to check the size of a PyArrow date32 type in terms of memory usage, you would look at the size of the integer it uses to represent dates. Since date32 uses a 32-bit integer, you would expect it to take up 4 bytes per value. 

If you're looking to confirm the size of the PyArrow date32 type in memory, you may try, eg.,  `pa.array`…`buffer().size`.

In summary, the `itemsize` attribute is not the correct attribute to check for the size of a `PyArrow` type within a `pd.ArrowDtype` object. Instead, you should refer to the PyArrow documentation or use PyArrow functions to determine the size of the type in memory.

**Comment by MarcoGorelli** [comment]
In summary, you shouldn't copy-and-paste AI generated answers on github issues

**Comment by longovin** [comment]
take

**Comment by echerrin** [comment]
take

**Comment by Charlie-H7** [comment]
I'm having a hard time trying to find what and where to find the relationship between numpy_dtype and the itemsize method. Going to the definition of itemsize method does not show an implementation of the method, so I'm not sure what it is doing. 

[pandas/pandas/core/dtypes/dtypes.py#L2213-L2216](https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/dtypes.py#L2213-L2216)

Any help is appreciated

**Comment by george-adams1** [comment]
take

**Comment by meet-vasita** [comment]
take

**Comment by mraabhijit** [comment]
Hi, I'd like to work on this issue. Is it available?

**Comment by meet-vasita** [comment]
Hi @mraabhijit, It's available you can work on it.

**Comment by mraabhijit** [comment]
The PR #62657 aims to address the current issue.
Although there are some unit tests which are failing. Not sure whether these are necessary for a successful merge or not.

The tests being:
- Numpy Dev
- python-dev (ubuntu-24.04)
- python-dev (macOS-13)
- python-dev (MacOS-14)
- python-dev (windows-2025)

Some guidance on how to address these would be much appreciated.
Thanks.

**Comment by mraabhijit** [comment]
@jorisvandenbossche seeking help with the CI failures here for the issue. 
I have tested the push on my local repo and the CI tests pass, however, the tests fail as mentioned above in this repo.
Some guidance would be helpful on how to address it.
